docs/src/main/asciidoc/extension-maturity-model.adoc

```diff
@@ -0,0 +1,89 @@
+////
+This guide is maintained in the main Quarkus repository
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
+////
+[id="extension-maturity-model"]
+= A maturity model for Quarkus extensions
+include::_attributes.adoc[]
+:diataxis-type: concept
+:categories: writing-extensions
+:topics: extensions
+:summary: Quarkus extensions can do a lot, or a little. This guide explains some of the capabilities extension authors might want to include.
+////
+The document header ends at the first blank line. Do not remove the blank line between the header and the abstract summary.
+////
+
+What makes a good Quarkus extension? What capabilities is a Quarkus extension expected to provide?
+
+image::extension-maturity-model.svg[A maturity model: Runtime actually works,Works in JVM mode,Works in dev mode, Works in native,Configuration support,Dev service,Dev UI, Use build-time application knowledge to eliminate wasteful runtime code paths (supersonic subatomic performance),Use build-time application knowledge to remove boilerplate,Codestart application template]
+
+The order in this model isn't exact. Different developers will have different views on what capabilities are most important. You may wish to (say) prioritise performance over enhancing your extension's Dev UI tile. That's fine!
+Also, not every step will apply to every extension. For example, you don't need a Dev Service if your extension doesn't depend on external services.
+It's completely OK to publish a first version of an extension that doesn't handle everything. In fact, it's OK if your extension _never_ gets to the more advanced features. This is a suggested pathway, not a minimum feature set.
+
+Also note that this list only includes the technical features of your extension.
+You might also want to think about how you share your extension, and how it presents itself to the world.
+The link:https://hub.quarkiverse.io/checklistfornewprojects/[new extension checklist] on the Quarkiverse Hub has a useful list of ways extensions can participate in the ecosystem.
+It's also a good idea to spend some time on the metadata in the link:https://quarkus.io/version/main/guides/extension-metadata#quarkus-extension-yaml[`quarkus-extension.yaml` file], which is used by Quarkus tooling.
+
+Here are some pointers on how to achieve those capabilities.
+
+== Runtime actually works
+
+=== Works in JVM mode
+
+For most extensions, this is the minimum expectation.
+When wrapping an existing library, this is usually trivial to achieve; if an extension is providing net-new capability, it might be a bit more work. Quarkus provides tools for xref:writing-extensions.adoc#testing-extensions[unit testing and integration testing] extensions.
+
```

Comment:
Here there is (at least) two categories:

- just wrapping the library
- exposing _managed_ clients

Reply:
What do we mean by a managed client in this context? Just injectable by CDI, or something more? Is that starting to get into the area of what I'd called programming model improvements elsewhere in the doc?